rowenc: Christmas cleanup

RaduBerinde opened the pull request:
This PR contains assorted cleanups of the row encoding code. This code has evolved organically and it is pretty hard to navigate all the details.

#### rowenc: reduce reliance on the entire TableDescriptor

This change reduces uses of the entire TableDescriptor.

Release note: None

#### descpb: add and use NoColumnID constant

This change adds a `descpb.NoColumnID` constant so we don't have to
use `descpb.ColumnID(encoding.NoColumnID)` everywhere.

Release note: None

#### rowenc: remove DatumAlloc.scratch

This commit removes the `DatumAlloc.scratch` field which is always
nil.

Release note: None

#### sql: move `DatumAlloc` to sem/tree

This commit moves `DatumAlloc` from `sql/rowenc` to `sql/sem/tree`.
This is a fairly low-level construct and is not used exclusively for
row encoding/decoding.

Release note: None

#### rowenc: move low-level key encoding functions to subpackage

The `rowenc` package contains a hefty mix of functions operating at
different conceptual levels - some use higher level constructs like
table and index descriptors, others are lower level utilities. The
naming of these functions doesn't help, for example `EncodeTableKey`
sounds like a top-level function but is actually a low level utility
that appends a single value to a key

This commit moves the lower level utilities for encoding datum values
into keys to the `rowenc/keyside` package.

Release note: None

#### rowenc: minor cleanup of array decoding code

The code around array decoding was confusing; we now have a single
`decodeArray` variant which is the inverse of `encodeArray`.

Release note: None

#### rowenc: move low-level value encoding functions to subpackage

This commit moves the lower level utilities for encoding datum values
into values to the `rowenc/valueside` package.

Release note: None

#### valueside: introduce ColumnIDDiff type

When encoding multiple values, we encode the differences between the
ColumnIDs. This difference is passed to `valueside.Encode`, but it is
passed as a `descpb.ColumnID`. This commit introduces a new type to
make this distinction more evident.

Release note: None

#### valueside: add Decoder helper

We have a few copies of the same logic of decoding multiple SQL values
from a Value, in particular in range feed handling code.

This commit adds a helper type that centralizes this logic, reducing
duplication.

Note that the query executione engines retain their own specialized
implementations.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74357)
<!-- Reviewable:end -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Very nice!

> Reviewed 23 of 23 files at r1, 15 of 15 files at r2, 2 of 2 files at r3, 135 of 135 files at r4, 53 of 53 files at r5, 2 of 2 files at r6, 55 of 55 files at r7, 23 of 23 files at r8, 11 of 11 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74357#-:-MsGIWrL9-Jl2nQGNh6G:b-kerl87)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @HonoreDB, @jordanlewis, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/rowenc/valueside/encode.go, line 105 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74357#-MsGHS1j-3nwNwe_PxTa:-MsGHS1j-3nwNwe_PxTb:bg9hvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f915c08e4e1a7ab8ac9c5162fa7517746d80e159/pkg/sql/rowenc/valueside/encode.go#L105)):*
> ```Go
> // columns are encoded in a single value, the difference relative to the
> // previous column ID is encoded for each column (to minimize space usage).
> type ColumnIDDiff uint32
> ```

super-nit: a nice alternate name could be `ColumnIDDelta`
___
*[pkg/sql/rowenc/valueside/encode.go, line 110 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74357#-MsGGTYO75Sc1MDEu51H:-MsGGTYO75Sc1MDEu51I:b-b5uy37) ([raw file](https://github.com/cockroachdb/cockroach/blob/f915c08e4e1a7ab8ac9c5162fa7517746d80e159/pkg/sql/rowenc/valueside/encode.go#L110)):*
> ```Go
> // This is used when we use value encodings not to write KV Values but other
> // purposes, for example transferring a value over DistSQL (in the row engine).
> const NoColumnID = ColumnIDDiff(encoding.NoColumnID)
> ```

nit: `NoColumnIDDiff`would make it more clear it's a ColumnIDDiff, not a ColumnID. Or maybe `UnspecifiedColumnIDDiff`?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR and Happy New Year!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74357#-:-MsLUR0o7dWLNPYDaoff:bewr9m8)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @dt, @HonoreDB, @jordanlewis, @mgartner, and @yuzefovich)

___
*[pkg/sql/rowenc/valueside/encode.go, line 105 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74357#-MsGHS1j-3nwNwe_PxTa:-MsIBSvECawBZzbsTEKK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f915c08e4e1a7ab8ac9c5162fa7517746d80e159/pkg/sql/rowenc/valueside/encode.go#L105)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

super-nit: a nice alternate name could be `ColumnIDDelta`
</blockquote></details>

Done.
___
*[pkg/sql/rowenc/valueside/encode.go, line 110 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74357#-MsGGTYO75Sc1MDEu51H:-MsIBTv8-67FuoBNwaML:brzboto) ([raw file](https://github.com/cockroachdb/cockroach/blob/f915c08e4e1a7ab8ac9c5162fa7517746d80e159/pkg/sql/rowenc/valueside/encode.go#L110)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: `NoColumnIDDiff`would make it more clear it's a ColumnIDDiff, not a ColumnID. Or maybe `UnspecifiedColumnIDDiff`?
</blockquote></details>

Meh, I left it, I think of it as the value for the "no column ID" situation. The type makes it clear what it is.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
rowenc: minor cleanup of array decoding code

The code around array decoding was confusing; we now have a single
`decodeArray` variant which is the inverse of `encodeArray`.

Release note: None

RaduBerinde committed:
valueside: add Decoder helper

We have a few copies of the same logic of decoding multiple SQL values
from a Value, in particular in range feed handling code.

This commit adds a helper type that centralizes this logic, reducing
duplication.

Note that the query executione engines retain their own specialized
implementations.

Release note: None

RaduBerinde committed:
valueside: introduce ColumnIDDelta type

When encoding multiple values, we encode the differences between the
ColumnIDs. This difference is passed to `valueside.Encode`, but it is
passed as a `descpb.ColumnID`. This commit introduces a new type to
make this distinction more evident.

Release note: None

RaduBerinde committed:
rowenc: move low-level key encoding functions to subpackage

The `rowenc` package contains a hefty mix of functions operating at
different conceptual levels - some use higher level constructs like
table and index descriptors, others are lower level utilities. The
naming of these functions doesn't help, for example `EncodeTableKey`
sounds like a top-level function but is actually a low level utility
that appends a single value to a key

This commit moves the lower level utilities for encoding datum values
into keys to the `rowenc/keyside` package.

Release note: None

RaduBerinde committed:
rowenc: move low-level value encoding functions to subpackage

This commit moves the lower level utilities for encoding datum values
into values to the `rowenc/valueside` package.

Release note: None

ajwerner commented on pkg/ccl/partitionccl/partition.go:
```diff
@@ -115,7 +115,7 @@ func valueEncodePartitionTuple(
 		}
 		value, err = rowenc.EncodeTableValue(
-			value, descpb.ColumnID(encoding.NoColumnID), datum, scratch,
+			value, descpb.NoColumnID, datum, scratch,
```
I've been thinking that these ids are getting annoying from a dependency hierarchy perspective. I hate the relationship to `tree` too. I don't know exactly where to slot it, but I want the ID types to be extremely low in the import tree and then, maybe, get forwarded up to `catalog` or something. What I have in mind is a `catid` package which can be imported by tree, descpb, etc. I'm running into this now because I want to move the new schema changer state (`scpb`) below `descpb` so it can be imported and am being thwarted by these IDs. I'll make that change after this goes in and would love to move `encoding.NoColumnID` to `catid.NoColumnID` or perhaps even `catid.NoColumn`. Thoughts?

ajwerner commented on pkg/sql/inverted/expression.go:
```diff
@@ -27,7 +27,7 @@ import (
 // If the inverted column stores an encoded datum, the encoding is
 // DatumEncoding_ASCENDING_KEY, and is performed using
-// EncodeTableKey(nil /* prefix */, val tree.Datum, encoding.Ascending).
+// rowside.Encode(nil /* prefix */, val tree.Datum, encoding.Ascending).
```
`rowside`?

ajwerner commented on pkg/sql/rowenc/keyside/doc.go:
```diff
@@ -0,0 +1,18 @@
+// descriptors.
+//
+// See also: docs/tech-notes/encoding.md.
+package keyside
```
why `keyside`, what's a `side`? `keyenc`?

ajwerner commented on pkg/sql/rowenc/valueside/doc.go:
```diff
@@ -0,0 +1,28 @@
+//    SQL values (columns) per roachpb.Value.
+//
+// See also: docs/tech-notes/encoding.md.
+package valueside
```
`rowside` is a bit clearer now in the context of `valueside`. I'm cool with the names but would also be cool with `keyenc`/`valenc`. I see that `valueenc` is frustrating with the `ee` in the middle.

ajwerner reviewed (approved):
LGTM

mgartner reviewed (approved):
> Reviewed 10 of 199 files at r10, 35 of 135 files at r13, 9 of 53 files at r14, 11 of 55 files at r16, 18 of 32 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74357)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @HonoreDB, @jordanlewis, @mgartner, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/rowenc/valueside/encode.go, line 110 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74357#-MsGGTYO75Sc1MDEu51H:-MsVQUg7B6pyz6UJCT8l:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f915c08e4e1a7ab8ac9c5162fa7517746d80e159/pkg/sql/rowenc/valueside/encode.go#L110)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Meh, I left it, I think of it as the value for the "no column ID" situation. The type makes it clear what it is.
</blockquote></details>

:+1: 


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74357)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @dt, @HonoreDB, @jordanlewis, @mgartner, @RaduBerinde, and @yuzefovich)

___
*[pkg/ccl/partitionccl/partition.go, line 118 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/74357#-MsDVpHE17tn7FRx2WvY-r11-118:-MsWyYta0k6TqErvYs-0:b57weex) ([raw file](https://github.com/cockroachdb/cockroach/blob/771ced9b9da5ed17a95125c3c868ebfbc5fd010c/pkg/ccl/partitionccl/partition.go#L118)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I've been thinking that these ids are getting annoying from a dependency hierarchy perspective. I hate the relationship to `tree` too. I don't know exactly where to slot it, but I want the ID types to be extremely low in the import tree and then, maybe, get forwarded up to `catalog` or something. What I have in mind is a `catid` package which can be imported by tree, descpb, etc. I'm running into this now because I want to move the new schema changer state (`scpb`) below `descpb` so it can be imported and am being thwarted by these IDs. I'll make that change after this goes in and would love to move `encoding.NoColumnID` to `catid.NoColumnID` or perhaps even `catid.NoColumn`. Thoughts?
</blockquote></details>

I agree they should be defined in a lower package I'd like the keyside/valueside code to not depend on the catalog or descpb.

Note though that the thing in `encoding` is not a simple column ID, it can also be a delta between column IDs. I introduce a type for that in `valueside` in the next-to-last commit in this PR.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
rowenc: remove DatumAlloc.scratch

This commit removes the `DatumAlloc.scratch` field which is always
nil.

Release note: None

RaduBerinde committed:
descpb: add and use NoColumnID constant

This change adds a `descpb.NoColumnID` constant so we don't have to
use `descpb.ColumnID(encoding.NoColumnID)` everywhere.

Release note: None

RaduBerinde committed:
rowenc: reduce reliance on the entire TableDescriptor

This change reduces uses of the entire TableDescriptor.

Release note: None

RaduBerinde committed:
sql: move `DatumAlloc` to sem/tree

This commit moves `DatumAlloc` from `sql/rowenc` to `sql/sem/tree`.
This is a fairly low-level construct and is not used exclusively for
row encoding/decoding.

Release note: None

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74357)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @dt, @HonoreDB, @jordanlewis, @mgartner, and @yuzefovich)

___
*[pkg/sql/inverted/expression.go, line 30 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/74357#-MsDVpHHRF6Zzg_ejzW1-r14-30:-MsqGuftDIIJM8gwQuqI:b-byiv19) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f40975534b1d2077dd894b3696943ce972b2991/pkg/sql/inverted/expression.go#L30)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`rowside`?
</blockquote></details>

Should be `keyside`, fixed.
___
*[pkg/sql/rowenc/valueside/doc.go, line 28 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/74357#-MsDVpHLQwVM0wCS2ylS-r16-28:-Msq9row4K_YhLOINlQG:b5ilcj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5456cbe7f7aa4be5db01701ff3d78bfa697eda84/pkg/sql/rowenc/valueside/doc.go#L28)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`rowside` is a bit clearer now in the context of `valueside`. I'm cool with the names but would also be cool with `keyenc`/`valenc`. I see that `valueenc` is frustrating with the `ee` in the middle. 
</blockquote></details>

I think rowenc/keyenc are better package names but not in actual usage, e.g. `rowside.Encode()` is better than `rowenc.Encode()` IMO.  That said, I don't have a problem switching if someone feels strongly (if not, I'll leave them to avoid reworking the commits).


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
